### PR TITLE
Fixing issues with workspace_key_prefix for s3 backend

### PR DESCRIPTION
**Pull request by rv-jmaggio**
I believe this will fix https://github.com/hashicorp/terraform/issues/16383 and https://github.com/hashicorp/terraform/issues/16387

### CODE REVIEW

**Comment by rv-jmaggio** [comment]
@jbardin @apparentlymart 
Ran into this bug today and I think I've got a fix here. I added some tests for this functionality and cleaned up a bit. However, I haven't manually tested this. I'm hoping this could be released in a minor provider version.

File: backend/remote-state/s3/backend_state.go [hunk 1]
```diff
@@ -51,6 +58,12 @@ func (b *Backend) keyEnv(key string) string {
 		return ""
 	}
 
+	parts = strings.Split(parts[1], "/")
```
Comment (jbardin) [inline backend/remote-state/s3/backend_state.go]: If the key names are allowed to have `/` characters, you're going to want `SplitN` with 3 here.

File: backend/remote-state/s3/backend_test.go [hunk 2]
```diff
@@ -249,6 +249,62 @@ func TestBackendExtraPaths(t *testing.T) {
...
+
+	if err := testGetWorkspaceForKey(b0, "tfstate", ""); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := testGetWorkspaceForKey(b0, "ws1/tfstate", "ws1"); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := testGetWorkspaceForKey(b1, "root/userA/ws1/tfstate", "ws1"); err != nil {
```
Comment (jbardin) [inline backend/remote-state/s3/backend_test.go]: If the prefix is `root/userA` and the key is `ws1/tfstate`, `ws1` can't be the workspace name.

File: backend/remote-state/s3/backend_test.go [hunk 3]
```diff
@@ -249,6 +249,62 @@ func TestBackendExtraPaths(t *testing.T) {
...
+		t.Fatal(err)
+	}
+
+	if err := testGetWorkspaceForKey(b1, "root/userA/ws2/tfstate", "ws2"); err != nil {
```
Comment (jbardin) [inline backend/remote-state/s3/backend_test.go]: Same as above for this check and the one below. The tests are checking the output of `getWorkspaceForKey`, but that doesn't match how it's used by the backend.

File: backend/remote-state/s3/backend_test.go [hunk 4]
```diff
@@ -249,6 +249,62 @@ func TestBackendExtraPaths(t *testing.T) {
...
+		t.Fatal(err)
+	}
+
+	if err := testGetWorkspaceForKey(b2, "root/userA/ws2/env1/tfstate", "ws2"); err != nil {
+		t.Fatal(err)
+	}
```
Comment (jbardin) [inline backend/remote-state/s3/backend_test.go]: Once of the new backend configs needs to be run through the TestBackend helper, which would have failed with either `b1` or `b2`.

**Comment by rv-jmaggio** [comment]
@jbardin I updated from your feedback. I did some testing in a playground and you are definitely correct on SplitN. Evidently I was a bit confused when I wrote my test cases, so I made them a bit more clear. I think this should be good to go now. I am happy to contribute to a tool I use nearly every day 😃

File: backend/remote-state/s3/backend_test.go [hunk 5]
```diff
@@ -249,6 +249,63 @@ func TestBackendExtraPaths(t *testing.T) {
 	}
 }
 
+func TestKeyEnv(t *testing.T) {
```
Comment (jbardin) [inline backend/remote-state/s3/backend_test.go]: Though the rest of the ACC tests pass again, this test still fails.

File: backend/remote-state/s3/backend_test.go [hunk 6]
```diff
@@ -249,6 +249,63 @@ func TestBackendExtraPaths(t *testing.T) {
...
+	}
+
+	if err := testGetWorkspaceForKey(b1, "project/env:/ws2/tfstate", "ws2"); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := testGetWorkspaceForKey(b2, "env:/ws3/tfstate", "ws3"); err != nil {
+		t.Fatal(err)
+	}
+
```
Comment (jbardin) [inline backend/remote-state/s3/backend_test.go]: You're not creating this bucket for the backend tests, though it currently fails earlier.

File: backend/remote-state/s3/backend_test.go [hunk 7]
```diff
@@ -249,6 +249,63 @@ func TestBackendExtraPaths(t *testing.T) {
 	}
 }
 
+func TestKeyEnv(t *testing.T) {
+	testACC(t)
+	bucketName := fmt.Sprintf("terraform-remote-s3-test-%x", time.Now().Unix())
+	keyName := "tfstate"
```
Comment (jbardin) [inline backend/remote-state/s3/backend_test.go]: These test cases are good, but let's go back to having a keyName with a `/` to make sure it's still supported after we introduce them into the prefix.

File: backend/remote-state/s3/backend_state.go [hunk 8]
```diff
@@ -25,24 +25,31 @@ func (b *Backend) States() ([]string, error) {
...
-// extract the env name from the S3 key
-func (b *Backend) keyEnv(key string) string {
-	// we have 3 parts, the prefix, the env name, and the key name
-	parts := strings.SplitN(key, "/", 3)
-	if len(parts) < 3 {
-		// no env here
+func getWorkspaceForKey(key string, b *Backend) string {
+	if b.workspaceKeyPrefix == "" {
+		parts := strings.Split(key, "/")
+		if len(parts) > 1 && parts[1] == key {
```
[the remark this answers is not in the document]
Comment (rv-jmaggio) [reply backend/remote-state/s3/backend_state.go]: Ah yes the intention was to check b.keyName, not key. Thanks

**Comment by rv-jmaggio** [comment]
@jbardin 
Ah, thanks for the pointer on `TF_ACC=1`.

So I made those changes, but I was getting an error when running `backend.TestBackend(t, b0, nil)`
```
--- FAIL: TestKeyEnv (23.36s)
	backend_test.go:339: creating S3 bucket terraform-remote-s3-test-5a39211b-0 in us-west-2
	backend_test.go:339: creating S3 bucket terraform-remote-s3-test-5a39211e-1 in us-west-2
	backend_test.go:339: creating S3 bucket terraform-remote-s3-test-5a392121-2 in us-west-2
	backend_test.go:308: bad: []string{"default"}
FAIL```

It seems like when workspace_key_prefix is set to `""` then the workspace folder `default` isn't being created? Not exactly sure what to make of this.

**Comment by rv-jmaggio** [comment]
Ah actually it seems to be related to the implementation for empty string workspace_key_prefix. I think I can get that fixed.

**Comment by rv-jmaggio** [comment]
I think there may be an issue with this implementation for a workspace_key_prefix set to `""`

https://github.com/hashicorp/terraform/blob/master/backend/remote-state/s3/backend_state.go#L180

**Comment by rv-jmaggio** [comment]
@jbardin I had to make 2 more changes to the backend_state implementation to allow for empty string prefix. I ran the acceptance tests (actually in acceptance test mode) and they all passed.

**Comment by jbardin** [comment]
Hi @rv-jmaggio,

Thanks for the update!
This is looking good. I'll do a final review after we return from the break and aim to get this into the next release.

**Pull request merged**